room.cpp

```diff
@@ -875,6 +877,69 @@ void Room::Private::updateDisplayname()
         emit q->displaynameChanged(q);
 }
 
+void Room::Private::toJson(QJsonObject &out) {
+    QJsonValue nowTimestamp { QDateTime::currentMSecsSinceEpoch() };
+    QJsonArray stateEvents;
+
+    QJsonObject nameEvent;
+    nameEvent["type"] = QString("m.room.name");
+    QJsonObject nameEventContent;
+    nameEventContent["name"] = this->name;
+    nameEvent["content"] = nameEventContent;
+    stateEvents.append(QJsonValue(nameEvent));
+
+    for (auto i : this->membersMap) {
+        QJsonObject content;
+        content["membership"] = QString("join");
+        content["displayname"] = i->displayname();
+        // avatar URL is not available
+
+        QJsonObject memberEvent;
+        memberEvent["type"] = QString("m.room.member");
+        memberEvent["sender"] = (i->id());
+        memberEvent["state_key"] = (i->id());
+        memberEvent["content"] = (content);
+        memberEvent["membership"] = QString("join");
+        memberEvent["origin_server_ts"] = nowTimestamp;
```

Comment:
Unfortunately, this won't work right. Consider what happens when the server has the clock running behind the client one: you'll miss events from your timeline. I might look for the newest event in the room timeline and take its `timestamp()` instead.

Reply:
The whole issue reminds me of necessity to store a pointer to the whole member event in the members map, rather than copy data from the event. I struggle to give a compelling example for join events but `sender` and `state_key` can easily be different for leave events (kicking/banning), let alone invite events (which are not well supported in libqmatrixclient so far). Subscribe to https://github.com/QMatrixClient/libqmatrixclient/issues/79 if interested.